gems/smithy/lib/smithy/views/client/plugin_list.rb

```diff
@@ -0,0 +1,84 @@
+# frozen_string_literal: true
+
+module Smithy
+  module Views
+    module Client
+      # @api private
+      class PluginList
+        include Enumerable
+
+        def initialize(plan)
+          @plan = plan
+          @plugins = default_plugins + transport_plugins('http') + generated_plugins
+          @plugins.each do |plugin|
+            require_path = plugin.require_path
+            next unless require_path
+
+            require_path = File.absolute_path(require_path) unless plugin.relative_path?
+            Kernel.require(require_path)
+          end
+        end
+
+        def each(&)
+          @plugins.each(&)
+        end
+
+        private
+
+        def namespace
+          Util::Namespace.namespace_from_gem_name(@plan.options[:gem_name])
+        end
+
+        def gem_name
+          @plan.options[:gem_name]
+        end
+
+        def gem_dir
+          @plan.options[:destination_root]
+        end
+
+        def default_plugins
+          Smithy::Client::Base.plugins.map do |plugin|
+            Plugin.new(class_name: plugin.name, require_path: nil, default: true)
+          end
+        end
+
+        # def weld_plugins
+        #   plugins = @plan.welds.map(&:plugins).reduce({}, :merge)
+        #   plugins.map do |class_name, path|
+        #     Plugin.new(class_name: class_name, path: path)
+        #   end
+        # end
+
+        def generated_plugins
+          # Nested namespaces for Plugins will not load unless we define them.
+          define_namespaces
+          plugins = {
+            "#{namespace}::Plugins::Endpoint" => "#{gem_dir}/lib/#{gem_name}/plugins/endpoint.rb"
+          }
+          plugins.map do |class_name, require_path|
+            Plugin.new(class_name: class_name, require_path: require_path)
+          end
+        end
+
+        def transport_plugins(protocol)
+          plugins = {
+            'http' => { 'Smithy::Client::Plugins::NetHTTP' => 'smithy-client/plugins/net_http' }
+          }[protocol]
+          plugins.map do |class_name, require_path|
+            Plugin.new(class_name: class_name, require_path: require_path, relative_path: true, requirable: true)
+          end
+        end
+
+        def define_namespaces
+          parent = Object
+          namespace.split('::') do |mod|
+            child = mod
+            parent.const_set(child, ::Module.new) unless parent.const_defined?(child)
```

Comment:
Is there a possible issue where we define something as a module that should be a class?

eg: `Module1::Class1::MyPlugin1` - we would define `Class1` as a module instead of a class.

I think for codegen...this shouldn't matter and I dont think it would break anything.

Reply:
Yeah, that is a limitation and would blow up. I'm not sure how else how to handle this. We could consider making documentation a post process? So after the gem is built, load the whole thing?

Reply:
I think its unlikely an issue - this would apply only for generated plugins, which we fully control and I think in general we won't/shouldn't have many generated plugins.  I think its also rare that you would want a plugin class to be defined in the namespace of another class instead of a module, so I'm not sure how much we should try to work around this.